test/websocket.test.js

```diff
@@ -2763,4 +2765,403 @@ describe('WebSocket', () => {
+          assert.ok(err instanceof RangeError);
+          assert.strictEqual(err.code, 'WS_ERR_INVALID_OPCODE');
+
+          // This should be reported as close with no close frame received:
+          ws.on('close', (code, reason) => {
+            assert.strictEqual(code, 1006);
+            assert.strictEqual(reason, '');
+            clientClosedCleanly = true;
+
+            if (serverClosedCleanly) {
+              wss.close(done);
+            }
+          });
+        });
+      });
+
+      wss.on('connection', (ws) => {
+        // Server's socket should never see an error (i.e. no ECONNRESET)
+        ws._socket.on('error', (err) => {
+          assert.fail(err);
+        });
+
+        // Server should report close due to protocol error close frame received:
+        ws.on('close', (code, reason) => {
+          assert.strictEqual(code, 1002);
+          assert.strictEqual(reason, '');
+          serverClosedCleanly = true;
+
+          if (clientClosedCleanly) {
+            wss.close(done);
+          }
+        });
+
+        // Write an invalid frame, triggering everything to break:
+        ws._socket.write(Buffer.from([0x85, 0x00]));
+      });
+    });
+
+    it('cleanly shuts down after simultaneous errors', (done) => {
+      let clientClosed = false;
+      let serverClosed = false;
+      let serverWs;
+
+      const wss = new WebSocket.Server({ port: 0 }, () => {
+        const ws = new WebSocket(`ws://localhost:${wss.address().port}`);
+
+        // Both peers will see an OPCODE error due to invalid frames:
+        ws.on('error', (err) => {
+          assert.strictEqual(err.code, 'WS_ERR_INVALID_OPCODE');
+        });
+
+        ws.on('close', () => {
+          clientClosed = true;
+          if (serverClosed) wss.close(done);
+        });
+
+        ws.on('open', () => {
+          // Write an invalid frame in both directions to trigger simultaneous failure:
+          serverWs._socket.write(Buffer.from([0x85, 0x00]));
+          ws._socket.write(Buffer.from([0x85, 0x00]));
+        });
+      });
+
+      wss.on('connection', (ws) => {
+        serverWs = ws;
+
+        // Both peers will see an OPCODE error due to invalid frames:
+        serverWs.on('error', (err) => {
+          assert.strictEqual(err.code, 'WS_ERR_INVALID_OPCODE');
+        });
+
+        serverWs.on('close', () => {
+          serverClosed = true;
+          if (clientClosed) wss.close(done);
+        });
+      });
+    });
+
+    it('handles an incoming close frame while compressing outgoing frames', (done) => {
+      const wss = new WebSocket.Server(
+        {
+          port: 0,
+          perMessageDeflate: { threshold: 0 }
+        },
+        () => {
+          const ws = new WebSocket(`ws://localhost:${wss.address().port}`, {
+            perMessageDeflate: { threshold: 0 }
+          });
+
+          ws.on('open', () => {
+            // Add 200ms delay to compression:
+            makeDeflateSlow(ws, 200);
```

Comment:
Can we avoid using timers? We can push the data directly to the socket without relying on the other peer, similarly to how it is done here https://github.com/websockets/ws/blob/145480a5b520ee951d848009d51069bfd7ed928c/test/create-websocket-stream.test.js#L445 and then assert that `sender._deflating === true`.